### PR TITLE
initial commit of gofuzz files

### PR DESCRIPTION
**Pull request by caffix**
moving the gofuzz files into tendermint core

### CODE REVIEW

File: p2p/gofuzz.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	remoteConn, _ = net.Pipe()
+}
+
+// Each execution of the fuzzer calls this function with random data provided.
+func Fuzz(data []byte) int {
+	var i int
+	datalen := len(data)
+
+	if datalen < 12 {
+		// small amounts of data will not due
```
Comment (odeke-em) [inline p2p/gofuzz.go]: Did you mean to say `will not do`?

File: p2p/gofuzz.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+		return -1
+	}
+
+	// simulate random remote peer
+	peer := randomPeer(data)
+	i += 6
```
Comment (odeke-em) [inline p2p/gofuzz.go]: This is the first instantiation of i, so perhaps could we make it:
```go
i := 6
```

File: p2p/gofuzz.go (hunk 1 above)
Comment (caffix) [reply p2p/gofuzz.go]: Yes. That is an error.

File: p2p/gofuzz.go (hunk 2 above)
Comment (caffix) [reply p2p/gofuzz.go]: Sounds good.

File: p2p/gofuzz.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+// randomPeer uses random bytes provided by the fuzzer to create a random peer.
+// TODO: the peer can be made more complete in order to support testing of other message types.
+func randomPeer(b []byte) *Peer {
+	privKey := crypto.GenPrivKeyEd25519()
+	config := DefaultPeerConfig()
+	config.AuthEnc = false
+	netAddr := randomAddr(b)
+
+	p, err := newInboundPeer(remoteConn, make(map[byte]Reactor),
+		make([]*ChannelDescriptor, 0), func(p *Peer, r interface{}) {}, privKey, config)
```
Comment (ebuchman) [inline p2p/gofuzz.go]: can a peer with no channels do anything ?

**Comment by ebuchman** [comment]
This is a good start to fuzzing the pexAddrsMessage, but I think we'll need to be more comprehensive.

Could we have the passed in `data` be well formed serialized messages themselves, and we build up a corpus of them? This Fuzz function feels a little too specific, and limitted in how much of the reactor.Receive it covers.

We should also look to fuzz the underlying connections recv routine too for more complete coverage of the peer package, independent of the pex_reactor

**Comment by ebuchman** [comment]
Ah, there's also lots of compile errors because of the recent refactor in the peer package to make `Peer` an interface

**Comment by ebuchman** [comment]
see https://github.com/tendermint/tendermint/pull/719

**Pull request closed without merging**